File: code/backend/Odin/src/main/kotlin/pt/isel/odin/service/interfaces/SubCategoryService.kt
package pt.isel.odin.service.interfaces

import pt.isel.odin.model.SubCategory

/**
 * Service for SubCategories.
 */
interface SubCategoryService {

    /**
     * Gets a SubCategory by its id.
     *
     * @param id the SubCategory id.
     */
    fun getById(id: Long): SubCategory

    /**
     * Gets all SubCategories.
     */
    fun getAll(): List<SubCategory>

    /**
     * Saves or Updates a SubCategory.
     *
     * @param subCategoryRequest the SubCategory to save.
     */
    fun save(subCategoryRequest: SubCategory): SubCategory

    /**
     * Deletes a SubCategory by its id.
     *
     * @param id the SubCategory id.
     */
    fun delete(id: Long)
}

File: code/frontend/Tyr/src/services/SubCategoryService.ts
import {makeApiRequest} from '../axios/apiRequest';
import {SubCategory} from "../model/SubCategory";

/**
 * Service to manage subcategories
 */
export class SubCategoryService {

    static basePath = '/subcategories';

    static async getById(id: number): Promise<SubCategory> {
        return makeApiRequest('get', `${(SubCategoryService.basePath)}/${id}`);
    }

    static async getAll(): Promise<SubCategory[]> {
        return makeApiRequest('get', SubCategoryService.basePath);
    }

    static async save(subCategoryRequest: SubCategory): Promise<SubCategory> {
        return makeApiRequest('post', `${(SubCategoryService.basePath)}/save`, subCategoryRequest);
    }

    static async update(subCategoryRequest: SubCategory): Promise<SubCategory> {
        return makeApiRequest('put', `${(SubCategoryService.basePath)}/update`, subCategoryRequest);
    }

    static async delete(id: number): Promise<void> {
        return makeApiRequest('delete', `${(SubCategoryService.basePath)}/${id}`);
    }
}